{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f256aff2-b4a3-4272-97a0-bd9e23c13a5b",
   "metadata": {},
   "source": [
    "# Scan for TF binding motifs and generate base GRN\n",
    "\n",
    "This notebook demonstrates how to scan for TF binding motifs. The base GRN will be generated by combining the ATAC-seq peaks and motif information.\n",
    "\n",
    "Much of this code is from CellOracle tutorial: https://morris-lab.github.io/CellOracle.documentation/notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_20200801.html\n",
    "\n",
    "### 0. Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20a69dbf-c0c1-4374-b914-f4140ac0edae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import os, sys, shutil, importlib, glob\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "%matplotlib inline\n",
    "\n",
    "plt.rcParams['figure.figsize'] = (15,7)\n",
    "plt.rcParams[\"savefig.dpi\"] = 600"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46a57586-52e3-4190-afa8-63281de3ee1d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "import celloracle as co\n",
    "from celloracle import motif_analysis as ma\n",
    "from celloracle.utility import save_as_pickled_object\n",
    "co.__version__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f859176a-174a-457f-9bec-1845298f5cc4",
   "metadata": {},
   "source": [
    "### 1. Reference genome data preparation\n",
    "Before starting the analysis, we need to make sure the reference genome data is installed with genomepy. If not, please install the correct reference genome using the instructions below.\n",
    "\n",
    "genomepy installs reference genome data under home directory. But if you have installed or want to install reference genome to another specific location, please specify the place using genomes_dir argument."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc2b1319-c6ef-4a9b-a856-3145cb5f63b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ref_genome = \"mm10\"\n",
    "# genomes_directory = \"SET GENOMES DIRECTORY HERE\"\n",
    "\n",
    "genome_installation = ma.is_genome_installed(ref_genome=ref_genome,\n",
    "                                             genomes_dir=genomes_directory)\n",
    "print(ref_genome, \"installation: \", genome_installation)\n",
    "\n",
    "if not genome_installation:\n",
    "    import genomepy\n",
    "    genomepy.install_genome(name=ref_genome, provider=\"UCSC\", genomes_dir=genomes_directory)\n",
    "else:\n",
    "    print(ref_genome, \"is installed.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e234522-eae8-4398-9c9c-7946cb438737",
   "metadata": {},
   "source": [
    "### 2. Load data\n",
    "In this notebook, we explain how to make a base GRN.\n",
    "\n",
    "Using the `processed_peak_file.csv` data file that was output from `6-tss-annotation.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4987b2a-1c83-44a3-b62c-773151a2afe9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_dir = \"data/base_grn_outputs/E9\"\n",
    "wt_or_ko = \"KO\"\n",
    "peaks = pd.read_csv(f\"{data_dir}/{wt_or_ko}_processed_peak_file.csv\", index_col=0)\n",
    "peaks.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "850881ac-3383-4bbb-9755-7a9f92225585",
   "metadata": {},
   "source": [
    "Here, the function below will check peak data format, including chromosome name and peak length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "add21ca9-9958-4e61-8d30-5e61ce406004",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "peaks = ma.check_peak_format(peaks, ref_genome, genomes_dir=genomes_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f981b96c-8cc1-4d18-aca3-ab24ad06e226",
   "metadata": {},
   "source": [
    "You can choose to use either a custom TF binding reference or CellOracle’s default motifs during the motif analysis. If you would like to use our default motifs, you can continue to the next step without loading any additional data. I am using the default!!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b98c4cd2-a816-41c1-be55-e059b454008c",
   "metadata": {},
   "source": [
    "### 3. Instantiate TFinfo object and search for TF binding motifs\n",
    "The motif analysis module has a custom class, TFinfo. The TFinfo object executes the steps below.\n",
    "\n",
    "1. Converts a peak data into a DNA sequences.\n",
    "2. Scans the DNA sequences searching for TF binding motifs.\n",
    "3. Post-processes the motif scan results.\n",
    "4. Converts data into appropriate format. You can convert data into base-GRN. The base GRN data can be formatted as either a python dictionary or pandas dataframe. This output will be the final base GRN used in the GRN model construction step.\n",
    "\n",
    "If your reference genome file are installed in non-default location, please speficy the location using genomes_dir.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e67ead-5608-4a5c-8f31-ed7540e2e6db",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Instantiate TFinfo object\n",
    "tfi = ma.TFinfo(peak_data_frame=peaks,\n",
    "                ref_genome=ref_genome,\n",
    "                genomes_dir=genomes_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d9226e7-63f5-4176-a908-886139d95400",
   "metadata": {},
   "source": [
    "You can specify the TF binding motif data as follows.\n",
    "\n",
    "tfi.scan(motifs=motifs)\n",
    "\n",
    "If you do not specify the motifs or set motifs to None, the default motifs will be loaded automatically.\n",
    "\n",
    "For mouse and human, “gimme.vertebrate.v5.0.” will be used as the default motifs.\n",
    "\n",
    "For another species, the species-specific TF binding motif data extracted from CisBP ver2.0 will be used.\n",
    "\n",
    "NEXT, scan for TF binding motifs. This will take a while!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77bbc872-9c84-4dfa-94f3-7b75c47a38ce",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "# Scan motifs. !!CAUTION!! This step may take several hours if you have many peaks!\n",
    "tfi.scan(fpr=0.02,\n",
    "         motifs=None,  # If you enter None, default motifs will be loaded.\n",
    "         verbose=True)\n",
    "\n",
    "# Save tfinfo object\n",
    "tfi.to_hdf5(file_path=f\"{data_dir}/{wt_or_ko}_test1.celloracle.tfinfo\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1508656e-10cc-4e18-9143-565773047656",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tfi.scanned_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a831f7a1-b314-4360-8e62-adcc48e2119c",
   "metadata": {},
   "source": [
    "### 4. Filtering motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8b9834a-4f57-4655-985d-4e599f20a43a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Reset filtering\n",
    "tfi.reset_filtering()\n",
    "\n",
    "# Do filtering\n",
    "tfi.filter_motifs_by_score(threshold=10)\n",
    "\n",
    "# Format post-filtering results.\n",
    "tfi.make_TFinfo_dataframe_and_dictionary(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79d2887f-a41d-4bd3-972a-5eeba9ab6fde",
   "metadata": {},
   "source": [
    "### 5. Get final base GRN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6c74acf-a61a-4e02-b3fa-594970a20943",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = tfi.to_dataframe()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3491b900-d255-490d-b2ce-5eb2e7c49891",
   "metadata": {},
   "source": [
    "### 6. Save results\n",
    "We will use this information when constructing the GRN models later. Save the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aedb2a0e-4495-4112-bc7b-ff59e0a08efc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Save result as a dataframe\n",
    "df = tfi.to_dataframe()\n",
    "df.to_parquet(f\"{data_dir}/{wt_or_ko}_base_GRN_dataframe.parquet\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
